<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Working with a real world data-set using SQL and Python

Estaimted time needed: **30** minutes

## Objectives

After complting this lab you will be able to:

*   Understand the dataset for Chicago Public School level performance
*   Store the dataset in SQLite database.
*   Retrieve metadata about tables and columns and query data from mixed case columns
*   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&download=true&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**:

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">link</a>.

Now review some of its contents.


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

##### The syntax for connecting to magic sql using sqllite is

**%sql sqlite://DatabaseName**

where DatabaseName will be your **.db** file


In [1]:
import csv, sqlite3

con = sqlite3.connect("RealWorldData.db")
cur = con.cursor()

In [ ]:
!pip install -q pandas==1.1.5

In [2]:
%load_ext sql

In [3]:
%sql sqlite:///RealWorldData.db

'Connected: @RealWorldData.db'

### Store the dataset in a Table

##### In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data

##### using SQL, it first needs to be stored in the database.

##### We will first read the csv files  from the given url  into pandas dataframes

##### Next we will be using the  df.to_sql() function to convert each csv file  to a table in sqlite  with the csv data loaded in it.


In [5]:
import pandas as pd

In [7]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01"
df = pd.read_csv(url)
df.head()

School_ID                                     NAME_OF_SCHOOL  \
0     610038                  Abraham Lincoln Elementary School   
1     610281  Adam Clayton Powell Paideia Community Academy ...   
2     610185                Adlai E Stevenson Elementary School   
3     609993                    Agustin Lara Elementary Academy   
4     610513                      Air Force Academy High School   

  Elementary, Middle, or High School         Street_Address     City State  \
0                                 ES        615 W Kemper Pl  Chicago    IL   
1                                 ES  7511 S South Shore Dr  Chicago    IL   
2                                 ES     8010 S Kostner Ave  Chicago    IL   
3                                 ES     4619 S Wolcott Ave  Chicago    IL   
4                                 HS        3630 S Wells St  Chicago    IL   

   ZIP_Code    Phone_Number  \
0     60614  (773) 534-5720   
1     60649  (773) 535-6650   
2     60652  (773) 535-2280   
3     60609  (773) 535-4389   
4     60609  (773) 535-1590   

                                                Link  \
0  http://schoolreports.cps.edu/SchoolProgressRep...   
1  http://schoolreports.cps.edu/SchoolProgressRep...   
2  http://schoolreports.cps.edu/SchoolProgressRep...   
3  http://schoolreports.cps.edu/SchoolProgressRep...   
4  http://schoolreports.cps.edu/SchoolProgressRep...   

                      Network_Manager  ... Freshman_on_Track_Rate__  \
0        Fullerton Elementary Network  ...                      NDA   
1           Skyway Elementary Network  ...                      NDA   
2           Midway Elementary Network  ...                      NDA   
3         Pershing Elementary Network  ...                      NDA   
4  Southwest Side High School Network  ...                     91.8   

  X_COORDINATE Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
0  1171699.458  1915829.428  41.924497 -87.644522                     7   
1  1196129.985  1856209.466  41.760324 -87.556736                    43   
2  1148427.165  1851012.215  41.747111 -87.731702                    70   
3  1164504.290  1873959.199  41.809757 -87.672145                    61   
4  1175177.622  1880745.126  41.828146 -87.632794                    34   

  COMMUNITY_AREA_NAME  Ward Police_District                     Location  
0        LINCOLN PARK    43              18  (41.92449696, -87.64452163)  
1         SOUTH SHORE     7               4  (41.76032435, -87.55673627)  
2             ASHBURN    13               8  (41.74711093, -87.73170248)  
3            NEW CITY    20               9    (41.8097569, -87.6721446)  
4       ARMOUR SQUARE    11               9  (41.82814609, -87.63279369)  

[5 rows x 78 columns]

In [10]:
df.to_sql("high_school",con,if_exists='replace', index=False,method="multi")

Double-click **here** for the solution.

<!--

import pandas
df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv")
df.to_sql("CHICAGO_CRIME_DATA", con, if_exists='replace', index=False, method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", con, if_exists='replace', index=False, method="multi")

-->


### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [39]:
# type in your query to retrieve list of all tables in the database
%sql select * from sqlite_master where type='table';

 * sqlite:///RealWorldData.db
Done.


[('table', 'high_school', 'high_school', 2, 'CREATE TABLE "high_school" (\n"School_ID" INTEGER,\n  "NAME_OF_SCHOOL" TEXT,\n  "Elementary, Middle, or High School" TEXT,\n  "Street_Address" TEXT,\ ... (2206 characters truncated) ... tude" REAL,\n  "COMMUNITY_AREA_NUMBER" INTEGER,\n  "COMMUNITY_AREA_NAME" TEXT,\n  "Ward" INTEGER,\n  "Police_District" INTEGER,\n  "Location" TEXT\n)')]

Double-click **here** for a hint

<!--
In sqlite the system catalog table called sqlite_master contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution using SQLite:

%sql SELECT name FROM sqlite_master WHERE type='table'

-->


### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [56]:
# type in your query to retrieve the number of columns in the SCHOOLS table
%sql PRAGMA table_info('high_school');

 * sqlite:///RealWorldData.db
Done.


[(0, 'School_ID', 'INTEGER', 0, None, 0),
 (1, 'NAME_OF_SCHOOL', 'TEXT', 0, None, 0),
 (2, 'Elementary, Middle, or High School', 'TEXT', 0, None, 0),
 (3, 'Street_Address', 'TEXT', 0, None, 0),
 (4, 'City', 'TEXT', 0, None, 0),
 (5, 'State', 'TEXT', 0, None, 0),
 (6, 'ZIP_Code', 'INTEGER', 0, None, 0),
 (7, 'Phone_Number', 'TEXT', 0, None, 0),
 (8, 'Link', 'TEXT', 0, None, 0),
 (9, 'Network_Manager', 'TEXT', 0, None, 0),
 (10, 'Collaborative_Name', 'TEXT', 0, None, 0),
 (11, 'Adequate_Yearly_Progress_Made_', 'TEXT', 0, None, 0),
 (12, 'Track_Schedule', 'TEXT', 0, None, 0),
 (13, 'CPS_Performance_Policy_Status', 'TEXT', 0, None, 0),
 (14, 'CPS_Performance_Policy_Level', 'TEXT', 0, None, 0),
 (15, 'HEALTHY_SCHOOL_CERTIFIED', 'TEXT', 0, None, 0),
 (16, 'Safety_Icon', 'TEXT', 0, None, 0),
 (17, 'SAFETY_SCORE', 'REAL', 0, None, 0),
 (18, 'Family_Involvement_Icon', 'TEXT', 0, None, 0),
 (19, 'Family_Involvement_Score', 'TEXT', 0, None, 0),
 (20, 'Environment_Icon', 'TEXT', 0, None, 0),
 (21, 'Environment_Score', 'REAL', 0, None, 0),
 (22, 'Instruction_Icon', 'TEXT', 0, None, 0),
 (23, 'Instruction_Score', 'REAL', 0, None, 0),
 (24, 'Leaders_Icon', 'TEXT', 0, None, 0),
 (25, 'Leaders_Score', 'TEXT', 0, None, 0),
 (26, 'Teachers_Icon', 'TEXT', 0, None, 0),
 (27, 'Teachers_Score', 'TEXT', 0, None, 0),
 (28, 'Parent_Engagement_Icon', 'TEXT', 0, None, 0),
 (29, 'Parent_Engagement_Score', 'TEXT', 0, None, 0),
 (30, 'Parent_Environment_Icon', 'TEXT', 0, None, 0),
 (31, 'Parent_Environment_Score', 'TEXT', 0, None, 0),
 (32, 'AVERAGE_STUDENT_ATTENDANCE', 'TEXT', 0, None, 0),
 (33, 'Rate_of_Misconducts__per_100_students_', 'REAL', 0, None, 0),
 (34, 'Average_Teacher_Attendance', 'TEXT', 0, None, 0),
 (35, 'Individualized_Education_Program_Compliance_Rate', 'TEXT', 0, None, 0),
 (36, 'Pk_2_Literacy__', 'TEXT', 0, None, 0),
 (37, 'Pk_2_Math__', 'TEXT', 0, None, 0),
 (38, 'Gr3_5_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (39, 'Gr3_5_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (40, 'Gr3_5_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (41, 'Gr3_5_Keep_Pace_Math__', 'TEXT', 0, None, 0),
 (42, 'Gr6_8_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (43, 'Gr6_8_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (44, 'Gr6_8_Keep_Pace_Math_', 'TEXT', 0, None, 0),
 (45, 'Gr6_8_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (46, 'Gr_8_Explore_Math__', 'TEXT', 0, None, 0),
 (47, 'Gr_8_Explore_Read__', 'TEXT', 0, None, 0),
 (48, 'ISAT_Exceeding_Math__', 'REAL', 0, None, 0),
 (49, 'ISAT_Exceeding_Reading__', 'REAL', 0, None, 0),
 (50, 'ISAT_Value_Add_Math', 'REAL', 0, None, 0),
 (51, 'ISAT_Value_Add_Read', 'REAL', 0, None, 0),
 (52, 'ISAT_Value_Add_Color_Math', 'TEXT', 0, None, 0),
 (53, 'ISAT_Value_Add_Color_Read', 'TEXT', 0, None, 0),
 (54, 'Students_Taking__Algebra__', 'TEXT', 0, None, 0),
 (55, 'Students_Passing__Algebra__', 'TEXT', 0, None, 0),
 (56, '9th Grade EXPLORE (2009)', 'TEXT', 0, None, 0),
 (57, '9th Grade EXPLORE (2010)', 'TEXT', 0, None, 0),
 (58, '10th Grade PLAN (2009)', 'TEXT', 0, None, 0),
 (59, '10th Grade PLAN (2010)', 'TEXT', 0, None, 0),
 (60, 'Net_Change_EXPLORE_and_PLAN', 'TEXT', 0, None, 0),
 (61, '11th Grade Average ACT (2011)', 'TEXT', 0, None, 0),
 (62, 'Net_Change_PLAN_and_ACT', 'TEXT', 0, None, 0),
 (63, 'College_Eligibility__', 'TEXT', 0, None, 0),
 (64, 'Graduation_Rate__', 'TEXT', 0, None, 0),
 (65, 'College_Enrollment_Rate__', 'TEXT', 0, None, 0),
 (66, 'COLLEGE_ENROLLMENT', 'INTEGER', 0, None, 0),
 (67, 'General_Services_Route', 'INTEGER', 0, None, 0),
 (68, 'Freshman_on_Track_Rate__', 'TEXT', 0, None, 0),
 (69, 'X_COORDINATE', 'REAL', 0, None, 0),
 (70, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (71, 'Latitude', 'REAL', 0, None, 0),
 (72, 'Longitude', 'REAL', 0, None, 0),
 (73, 'COMMUNITY_AREA_NUMBER', 'INTEGER', 0, None, 0),
 (74, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (75, 'Ward', 'INTEGER', 0, None, 0),
 (76, 'Police_District', 'INTEGER', 0, None, 0),
 (77, 'Location', 'TEXT', 0, None, 0)]

Double-click **here** for the solution.

<!-- Solution SQLite:
%sql SELECT name FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->


Now retrieve the list of columns in SCHOOLS table and their column type (datatype) and length.


In [65]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length
%sql SELECT name, typeof(name), length(name) FROM PRAGMA_TABLE_INFO('high_school');

 * sqlite:///RealWorldData.db
Done.


[('School_ID', 'text', 9),
 ('NAME_OF_SCHOOL', 'text', 14),
 ('Elementary, Middle, or High School', 'text', 34),
 ('Street_Address', 'text', 14),
 ('City', 'text', 4),
 ('State', 'text', 5),
 ('ZIP_Code', 'text', 8),
 ('Phone_Number', 'text', 12),
 ('Link', 'text', 4),
 ('Network_Manager', 'text', 15),
 ('Collaborative_Name', 'text', 18),
 ('Adequate_Yearly_Progress_Made_', 'text', 30),
 ('Track_Schedule', 'text', 14),
 ('CPS_Performance_Policy_Status', 'text', 29),
 ('CPS_Performance_Policy_Level', 'text', 28),
 ('HEALTHY_SCHOOL_CERTIFIED', 'text', 24),
 ('Safety_Icon', 'text', 11),
 ('SAFETY_SCORE', 'text', 12),
 ('Family_Involvement_Icon', 'text', 23),
 ('Family_Involvement_Score', 'text', 24),
 ('Environment_Icon', 'text', 16),
 ('Environment_Score', 'text', 17),
 ('Instruction_Icon', 'text', 16),
 ('Instruction_Score', 'text', 17),
 ('Leaders_Icon', 'text', 12),
 ('Leaders_Score', 'text', 13),
 ('Teachers_Icon', 'text', 13),
 ('Teachers_Score', 'text', 14),
 ('Parent_Engagement_Icon', 'text', 22),
 ('Parent_Engagement_Score', 'text', 23),
 ('Parent_Environment_Icon', 'text', 23),
 ('Parent_Environment_Score', 'text', 24),
 ('AVERAGE_STUDENT_ATTENDANCE', 'text', 26),
 ('Rate_of_Misconducts__per_100_students_', 'text', 38),
 ('Average_Teacher_Attendance', 'text', 26),
 ('Individualized_Education_Program_Compliance_Rate', 'text', 48),
 ('Pk_2_Literacy__', 'text', 15),
 ('Pk_2_Math__', 'text', 11),
 ('Gr3_5_Grade_Level_Math__', 'text', 24),
 ('Gr3_5_Grade_Level_Read__', 'text', 24),
 ('Gr3_5_Keep_Pace_Read__', 'text', 22),
 ('Gr3_5_Keep_Pace_Math__', 'text', 22),
 ('Gr6_8_Grade_Level_Math__', 'text', 24),
 ('Gr6_8_Grade_Level_Read__', 'text', 24),
 ('Gr6_8_Keep_Pace_Math_', 'text', 21),
 ('Gr6_8_Keep_Pace_Read__', 'text', 22),
 ('Gr_8_Explore_Math__', 'text', 19),
 ('Gr_8_Explore_Read__', 'text', 19),
 ('ISAT_Exceeding_Math__', 'text', 21),
 ('ISAT_Exceeding_Reading__', 'text', 24),
 ('ISAT_Value_Add_Math', 'text', 19),
 ('ISAT_Value_Add_Read', 'text', 19),
 ('ISAT_Value_Add_Color_Math', 'text', 25),
 ('ISAT_Value_Add_Color_Read', 'text', 25),
 ('Students_Taking__Algebra__', 'text', 26),
 ('Students_Passing__Algebra__', 'text', 27),
 ('9th Grade EXPLORE (2009)', 'text', 24),
 ('9th Grade EXPLORE (2010)', 'text', 24),
 ('10th Grade PLAN (2009)', 'text', 22),
 ('10th Grade PLAN (2010)', 'text', 22),
 ('Net_Change_EXPLORE_and_PLAN', 'text', 27),
 ('11th Grade Average ACT (2011)', 'text', 29),
 ('Net_Change_PLAN_and_ACT', 'text', 23),
 ('College_Eligibility__', 'text', 21),
 ('Graduation_Rate__', 'text', 17),
 ('College_Enrollment_Rate__', 'text', 25),
 ('COLLEGE_ENROLLMENT', 'text', 18),
 ('General_Services_Route', 'text', 22),
 ('Freshman_on_Track_Rate__', 'text', 24),
 ('X_COORDINATE', 'text', 12),
 ('Y_COORDINATE', 'text', 12),
 ('Latitude', 'text', 8),
 ('Longitude', 'text', 9),
 ('COMMUNITY_AREA_NUMBER', 'text', 21),
 ('COMMUNITY_AREA_NAME', 'text', 19),
 ('Ward', 'text', 4),
 ('Police_District', 'text', 15),
 ('Location', 'text', 8)]

Double-click **here** for the solution.

<!-- Solution 

%sql SELECT name,typeof(name),length(name) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->


### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "\_"?


## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [81]:
%sql select count(*) from high_school where "Elementary, Middle, or High School"="ES";

 * sqlite:///RealWorldData.db
Done.


[(462,)]

Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from CHICAGO_PUBLIC_SCHOOLS_DATA where "Elementary, Middle, or High School"='ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [84]:
%sql select max(SAFETY_SCORE) from high_school;

 * sqlite:///RealWorldData.db
Done.


[(99.0,)]

Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from CHICAGO_PUBLIC_SCHOOLS_DATA

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [88]:
%sql select NAME_OF_SCHOOL,SAFETY_SCORE from high_school where SAFETY_SCORE=(select max(SAFETY_SCORE) from high_school);

 * sqlite:///RealWorldData.db
Done.


[('Abraham Lincoln Elementary School', 99.0),
 ('Alexander Graham Bell Elementary School', 99.0),
 ('Annie Keller Elementary Gifted Magnet School', 99.0),
 ('Augustus H Burley Elementary School', 99.0),
 ('Edgar Allan Poe Elementary Classical School', 99.0),
 ('Edgebrook Elementary School', 99.0),
 ('Ellen Mitchell Elementary School', 99.0),
 ('James E McDade Elementary Classical School', 99.0),
 ('James G Blaine Elementary School', 99.0),
 ('LaSalle Elementary Language Academy', 99.0),
 ('Mary E Courtenay Elementary Language Arts Center', 99.0),
 ('Northside College Preparatory High School', 99.0),
 ('Northside Learning Center High School', 99.0),
 ('Norwood Park Elementary School', 99.0),
 ('Oriole Park Elementary School', 99.0),
 ('Sauganash Elementary School', 99.0),
 ('Stephen Decatur Classical Elementary School', 99.0),
 ('Talman Elementary School', 99.0),
 ('Wildwood Elementary School', 99.0)]

Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from SCHOOLS where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where \
  Safety_Score= (select MAX(Safety_Score) from CHICAGO_PUBLIC_SCHOOLS_DATA)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [100]:
%sql select NAME_OF_SCHOOL,"AVERAGE_STUDENT_ATTENDANCE" from high_school order by "AVERAGE_STUDENT_ATTENDANCE" desc limit 10;

 * sqlite:///RealWorldData.db
Done.


[('John Charles Haines Elementary School', '98.40%'),
 ('James Ward Elementary School', '97.80%'),
 ('Edgar Allan Poe Elementary Classical School', '97.60%'),
 ('Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 ('Rachel Carson Elementary School', '97.60%'),
 ('Annie Keller Elementary Gifted Magnet School', '97.50%'),
 ('Andrew Jackson Elementary Language Academy', '97.40%'),
 ('Lenart Elementary Regional Gifted Center', '97.40%'),
 ('Disney II Magnet School', '97.30%'),
 ('John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from CHICAGO_PUBLIC_SCHOOLS_DATA \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [101]:
%sql select NAME_OF_SCHOOL,"AVERAGE_STUDENT_ATTENDANCE" from high_school order by "AVERAGE_STUDENT_ATTENDANCE" limit 5;

 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [103]:
%sql select NAME_OF_SCHOOL,replace("AVERAGE_STUDENT_ATTENDANCE","%","") as attendance from high_school order by "AVERAGE_STUDENT_ATTENDANCE" limit 5;

 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Dyett High School', '62.50'),
 ('Wendell Phillips Academy High School', '63.00')]

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''

-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [109]:
%%sql 
select NAME_OF_SCHOOL,replace("AVERAGE_STUDENT_ATTENDANCE","%","") as attendance 
from high_school 
where cast(replace("AVERAGE_STUDENT_ATTENDANCE","%","") as double)<70;

 * sqlite:///RealWorldData.db
Done.


[('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Chicago Vocational Career Academy High School', '68.80'),
 ('Dyett High School', '62.50'),
 ('Manley Career Academy High School', '66.80'),
 ('Orr Academy High School', '66.30'),
 ('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Roberto Clemente Community Academy High School', '69.60'),
 ('Wendell Phillips Academy High School', '63.00')]

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)

-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [111]:
%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from high_school \
   group by Community_Area_Name 

 * sqlite:///RealWorldData.db
Done.


[('ALBANY PARK', 6864),
 ('ARCHER HEIGHTS', 4823),
 ('ARMOUR SQUARE', 1458),
 ('ASHBURN', 6483),
 ('AUBURN GRESHAM', 4175),
 ('AUSTIN', 10933),
 ('AVALON PARK', 1522),
 ('AVONDALE', 3640),
 ('BELMONT CRAGIN', 14386),
 ('BEVERLY', 1636),
 ('BRIDGEPORT', 3167),
 ('BRIGHTON PARK', 9647),
 ('BURNSIDE', 549),
 ('CALUMET HEIGHTS', 1568),
 ('CHATHAM', 5042),
 ('CHICAGO LAWN', 7086),
 ('CLEARING', 2085),
 ('DOUGLAS', 4670),
 ('DUNNING', 4568),
 ('EAST GARFIELD PARK', 5337),
 ('EAST SIDE', 5305),
 ('EDGEWATER', 4600),
 ('EDISON PARK', 910),
 ('ENGLEWOOD', 6832),
 ('FOREST GLEN', 1431),
 ('FULLER PARK', 531),
 ('GAGE PARK', 9915),
 ('GARFIELD RIDGE', 4552),
 ('GRAND BOULEVARD', 2809),
 ('GREATER GRAND CROSSING', 4051),
 ('HEGEWISCH', 963),
 ('HERMOSA', 3975),
 ('HUMBOLDT PARK', 8620),
 ('HYDE PARK', 1930),
 ('IRVING PARK', 7764),
 ('JEFFERSON PARK', 1755),
 ('KENWOOD', 4287),
 ('LAKE VIEW', 7055),
 ('LINCOLN PARK', 5615),
 ('LINCOLN SQUARE', 4132),
 ('LOGAN SQUARE', 7351),
 ('LOOP', 871),
 ('LOWER WEST SIDE', 7257),
 ('MCKINLEY PARK', 1552),
 ('MONTCLARE', 1317),
 ('MORGAN PARK', 3271),
 ('MOUNT GREENWOOD', 2091),
 ('NEAR NORTH SIDE', 3362),
 ('NEAR SOUTH SIDE', 1378),
 ('NEAR WEST SIDE', 7975),
 ('NEW CITY', 7922),
 ('NORTH CENTER', 7541),
 ('NORTH LAWNDALE', 5146),
 ('NORTH PARK', 4210),
 ('NORWOOD PARK', 6469),
 ('OAKLAND', 140),
 ('OHARE', 786),
 ('PORTAGE PARK', 6954),
 ('PULLMAN', 1620),
 ('RIVERDALE', 1547),
 ('ROGERS PARK', 4068),
 ('ROSELAND', 7020),
 ('SOUTH CHICAGO', 4043),
 ('SOUTH DEERING', 1859),
 ('SOUTH LAWNDALE', 14793),
 ('SOUTH SHORE', 4543),
 ('UPTOWN', 4388),
 ('WASHINGTON HEIGHTS', 4006),
 ('WASHINGTON PARK', 2648),
 ('WEST ELSDON', 3700),
 ('WEST ENGLEWOOD', 5946),
 ('WEST GARFIELD PARK', 2622),
 ('WEST LAWN', 4207),
 ('WEST PULLMAN', 3240),
 ('WEST RIDGE', 8197),
 ('WEST TOWN', 9429),
 ('WOODLAWN', 4206)]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [116]:
%sql select COMMUNITY_AREA_NAME from \
 (select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
  from high_school \
  group by COMMUNITY_AREA_NAME )\
 order by TOTAL_ENROLLMENT\
 limit 5;

 * sqlite:///RealWorldData.db
Done.


[('OAKLAND',), ('FULLER PARK',), ('BURNSIDE',), ('OHARE',), ('LOOP',)]

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   LIMIT 5 

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [129]:
%sql select NAME_OF_SCHOOL, SAFETY_SCORE from high_school where SAFETY_SCORE!= 'None' order by SAFETY_SCORE  limit 5;

 * sqlite:///RealWorldData.db
Done.


[('Edmond Burke Elementary School', 1.0),
 ("Luke O'Toole Elementary School", 5.0),
 ('George W Tilton Elementary School', 6.0),
 ('Foster Park Elementary School', 11.0),
 ('Emil G Hirsch Metropolitan High School', 13.0)]

Double-click **here** for the solution.

<!-- Solution SQLite:

%sql SELECT name_of_school, safety_score \
FROM CHICAGO_PUBLIC_SCHOOLS_DATA  where safety_score !='None' \
ORDER BY safety_score \
LIMIT 5

-->


### Problem 11

##### Get the hardship index for the community area which has College Enrollment of 4368


In [132]:
%sql select * from high_school where COLLEGE_ENROLLMENT=4368;

 * sqlite:///RealWorldData.db
Done.


[(609720, 'Albert G Lane Technical High School', 'HS', '2501 W Addison St', 'Chicago', 'IL', 60618, '(773) 534-5400', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609720.pdf', 'North-Northwest Side High School Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'Yes', 'Standard', 'Not on Probation', 'Level 1', 'No', 'Very Strong', 88.0, 'NDA', 'NDA', 'Strong', 62.0, 'Average', 52.0, 'Weak', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', '96.30%', 2.1, '96.20%', '99.40%', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', None, None, None, None, 'NDA', 'NDA', 'NDA', 'NDA', '19.1', '19.5', '19.9', '20.1', '1', '23.4', '3.5', '67.9', '92.2', '79.8', 4368, 35, '90.7', 1158975.392, 1923791.705, 41.94661693, -87.69105603, 5, 'NORTH CENTER', 47, 19, '(41.94661693, -87.69105603)')]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index from CENSUS_DATA CD, CHICAGO_PUBLIC_SCHOOLS_DATA CPS 
where CD.community_area_number = CPS.community_area_number 
and college_enrollment = 4368


-->


### Problem 12

##### Get the hardship index for the community area which has the highest value for College Enrollment


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select community_area_number, community_area_name, hardship_index from CENSUS_DATA \
   where community_area_number in \
   ( select community_area_number from CHICAGO_PUBLIC_SCHOOLS_DATA order by college_enrollment desc limit 1 )
-->


## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Rav Ahuja</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                        |
| ----------------- | ------- | ----------------- | ----------------------------------------- |
| 2022-03-04        | 2.2     | Lakshmi Holla     | Made changes in markdown cells            |
| 2020-11-27        | 2.1     | Sannareddy Ramesh | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya           | Moved lab to course repo in GitLab        |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
